### DISCUSSION TITLE
Does a lua script persist state across invocations?

### DISCUSSION THREAD

**Discussion by brunson** [fluent/fluent-bit#5682] [Q&A]
Hi,

I'd like to implement some per tag metrics and I can't find whether a lua script can maintain state across separate invocations. Is the execution environment a blank slate for every log event a script handles or could I create an array of counters that are available on successive events?

Thank you for your help!

**Comment by edsiper** [comment]
by design it does not keep a state between iterations.

cc: @tarruda

**Reply by tarruda** [reply]
@edsiper we never reset the lua state, so yes, state is shared between log lines. 
It can be useful keep state as shown in this xml parsing example: https://github.com/calyptia/fluent-bit-lua-samples/tree/main/xml-parsing

**Comment by brunson** [comment]
Thank you both for your quick responses.

Sincerely,
e.
